server/src/inmantals/server.py

```diff
@@ -300,27 +462,84 @@ async def initialized(self):
 
     async def shutdown(self, **kwargs):
         self.shutdown_requested = True
-        self.cleanup_tmp()
+        self._cleanup_tmp_projects()
         self.threadpool.shutdown(cancel_futures=True)
 
-    def cleanup_tmp(self):
-        self.project_dir.cleanup()
+    def _cleanup_tmp_projects(self):
+        for workspace in self._workspace_folders.values():
+            workspace.cleanup()
 
     async def exit(self, **kwargs):
         self.running = False
 
     async def textDocument_didOpen(self, **kwargs):  # noqa: N802
         pass
 
+    async def window_onDidChangeTextEditorViewColumn(self, **kwargs):  # noqa: N802
+        pass
+
+    async def window_onDidChangeActiveTextEditor(self, **kwargs):  # noqa: N802
+        pass
+
     async def textDocument_didChange(self, **kwargs):  # noqa: N802
         pass
 
     async def textDocument_didSave(self, **kwargs):  # noqa: N802
-        await self.compile_and_anchor()
+        logger.info(f"document saved, should probably do something here {kwargs=}")
+        # {'textDocument': {'uri': 'file:///home/hugo/tmp/tmp_module/test-module-v2/model/_init.cf'}}
+        try:
+            # r = await self.dispatch_method(1, "getWorkspaceFolder", file_uri)
+            # await self.send_notification("textDocument/publishDiagnostics", publish_params.dict())
+            # path = URI(file_uri).path
+            file_uri: URI = kwargs["textDocument"]["uri"]
+            folder = await self.send_notification("workspace/getWorkspaceFolder", {"uri": file_uri})
+
+            # logger.info(f"{self.anchormap.keys()=}")
+            # logger.info(f"{self.reverse_anchormap.keys()=}")
+            logger.debug(f"{folder}")
+
+            # folder = self.getWorkspaceFolder(file_uri)
+        except KeyError as e:
+            logger.debug(f"{e}")
+            folder = None
+
+        await self.compile_and_anchor([folder])
 
     async def textDocument_didClose(self, **kwargs):  # noqa: N802
         pass
 
+    async def workspace_DidChangeConfiguration(self, **kwargs):  # noqa: N802
+        logger.info(f"workspace_DidChangeConfiguration, should probably do something HERE {kwargs=}")
+        await self.compile_and_anchor()
```

Comment:
This needs to change and currently any change to the settings causes the server to crash and I can't figure out why

Reply:
It looks like the problem arises here: https://github.com/inmanta/vscode-inmanta/blob/0fb06b8861f200c297ed426d2d794e389f64a3af/src/extension.ts#L327
Because we restart the ls when the workspace configuration changes here: https://github.com/inmanta/vscode-inmanta/blob/0fb06b8861f200c297ed426d2d794e389f64a3af/src/extension.ts#L311-L315
I did get this trace:
```
rejected promise not handled within 1 second: Error: dictionary changed size during iteration
extensionHostProcess.js:100
stack trace: Error: dictionary changed size during iteration
	at handleResponse (/home/hugo/work/inmanta/github-repos/vscode-inmanta/node_modules/vscode-jsonrpc/lib/common/connection.js:567:48)
	at handleMessage (/home/hugo/work/inmanta/github-repos/vscode-inmanta/node_modules/vscode-jsonrpc/lib/common/connection.js:345:13)
	at processMessageQueue (/home/hugo/work/inmanta/github-repos/vscode-inmanta/node_modules/vscode-jsonrpc/lib/common/connection.js:362:17)
	at Immediate.<anonymous> (/home/hugo/work/inmanta/github-repos/vscode-inmanta/node_modules/vscode-jsonrpc/lib/common/connection.js:334:13)
	at process.processImmediate (node:internal/timers:466:21)
	at process.callbackTrampoline (node:internal/async_hooks:130:17)
```

But it is as far as I managed to get